src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicSession.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Security;
+using System.Text;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal
+{
+    internal sealed class MsQuicSession : IDisposable
+    {
+        private bool _disposed = false;
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _api;
+        private bool _opened;
+
+        internal MsQuicSession(string registration)
+        {
+            _api = new MsQuicApi();
+
+            _api.RegistrationOpen(Encoding.ASCII.GetBytes(registration));
+        }
+
+        public MsQuicConnection ConnectionOpen(IPEndPoint endpoint, SslClientAuthenticationOptions sslClientAuthenticationOptions, IPEndPoint localEndpoint)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslClientAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                SetPeerBiDirectionalStreamCount(100);
+                SetPeerUnidirectionalStreamCount(100);
+            }
+
+            uint status = _api.ConnectionOpenDelegate(
+                _nativeObjPtr,
+                MsQuicConnection.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr connectionPtr);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return new MsQuicConnection(endpoint, _api, connectionPtr, sslClientAuthenticationOptions);
+        }
+
+        public MsQuicListener ListenerOpen(IPEndPoint listenEndPoint, SslServerAuthenticationOptions sslServerAuthenticationOptions)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslServerAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                SetPeerBiDirectionalStreamCount(100);
```

Comment:
configurable?

Reply:
What is the purpose of having a limit at all?

Reply:
It's a type of flow control. The number of streams you're willing to accept and buffer data for. If you make this too big, an attacker can trivially make you allocate a TON of memory, with a single small packet. That's one of the reasons msquic limits this to 64K, though the protocol allows for 2^62.